_posts/2018-05-31-regular-types.md

```diff
@@ -0,0 +1,874 @@
+location?" This is an easy predicate to write in C++, but nonsense for a
+traditional logician. In general, we focus on predicates that observe "the
+value" rather than the identity of the instance.
+
+Stepanov goes on to describe the built-in notion of equality for most types:
+bitwise equality for ints and pointers. Floating point values are glossed over a
+bit via "although there are sometimes minor deviations like distinct positive
+and negative zero representations." From this as a basis, we can begin to build
+up a definition of equality for aggregates, but immediately get into trouble
+because of heap allocations. "... objects which are naturally variable sized
+must be constructed in C++ out of multiple simple structs, connected by
+pointers. In such cases, we say that the object has **remote** parts. For such
+objects, the equality operator must compare the remote parts ..."
+
+This reasoning continues, including some discussion about the physical state vs.
+logical state of a type (for instance: the capacity of a vector does not figure
+into its equality comparison, nor does the use of SSO affect equality comparison
+for string).
+
+Ultimately, Stepanov proposes the following definition, "although it still
+leaves room for judgement":
+
+**"Definition**: two objects are equal if their corresponding parts are equal
+(applied recursively), including remote parts (but not comparing their
+addresses), excluding inessential components, and excluding components which
+identify related objects."
+
+If we have a solid understanding of how to compare two instances of our type (by
+value, focusing on the logical state, not comparing by identity/memory
+location), and if our type implements all the syntactic/semantic requirements
+for Regular then we have implemented a Regular type. That should always be our
+starting point when designing a value type.
+
+## Modern Regular Types and const
+
+With an agreed-upon understanding of Regular, we can start to examine the ways
+that Regular is used. We can reasonably assume that a Regular type can be fed
+through standard algorithms, since that's a primary motivation for defining
+Regularity (or, conversely, the implied requirement for the algorithms, although
+the legacy algorithms are rarely specified in terms of semantic requirements).
+
+So we can certainly express that this is safe and correct for Regular types,
+yes?
+
+```c++
+void DoSomething(const T& t);  // May be any valid C++
+
+const T a = SomeT();
+const T b = SomeT();
+if (a == b) {
+  DoSomething(a);
+  assert(a == b);  // Note: we're assuming the semantics ==
+}
+```
+
+This seems straightforward: we have two const values. If they are equal, it
+doesn't matter what operation we perform on one of them, they will remain equal.
+It also doesn't matter if we modify any other global state. It does imply one
+additional requirement, however: the normal semantics of `const` must be
```

Comment:
additional requirement beyond Regular: the normal ...

Reply:
Done